### DISCUSSION TITLE
smx509 包代码打磨

### DISCUSSION THREAD

**Discussion by xuyang2** [emmansun/gmsm#21] [General]
比如目标是：
尽量与标准库 x509 包的代码差异最小，方便以后同步代码，
还是：
尽量复用标准库 x509 包，使 smx509 包的代码尽量减少？

以 `smx509.ParsePKIXPublicKey` 为例，现在是把 oidNamedCurveP256SM2 摘出来单独处理，
其余情况调用 `x509.ParsePKIXPublicKey()`

但因为 `smx509.namedCurveFromOID` 已经添加支持了 `oidNamedCurveP256SM2`
所以把 `x509.ParsePKIXPublicKey`  方法体整个复制过来，应该也是可以的。

**Comment by emmansun** [comment]
确实，处在golang语言SDK之外，又要尽量保持一致，同时可维护性还要高一点。还有一个主要问题是支持、兼容不同版本的golang SDK。这也是我不想动手改造支持国密tls 1.3的主要原因：改造维护成本太高。

**Comment by xuyang2** [comment]
除了 TLS1.3 (RFC8998) ，更加麻烦的是国密 TLCP (GB/T 38636-2020) 
TLCP 是基于 TLS 1.1，并参考了 TLS 1.2 的安全特性

而且 TLS1.3 (RFC8998) 即使支持了，估计也用处不大。
国内推的是双证书体系，RFC8998 搞成单证了，大概率不会被国内采纳为标准。

**Reply by emmansun** [reply]
TLCP (GB/T 38636-2020)这个在国内应用广泛吗？或者应用前景广泛吗？

**Reply by xuyang2** [reply]
- [https://catbro666.github.io/posts/59c71edb/](https://github.com/catbro666/catbro666.github.io/blob/a19f4e1548fe33fb0048b6b6132a26949a6cec8e/posts/59c71edb/index.html#L395)
- https://www.v2ex.com/t/720857?p=1
- https://wenku.baidu.com/view/c25b6b74a417866fb84a8e72
- https://github.com/lzimd/crypt-standard
- https://mp.weixin.qq.com/s/0aM9tL9PdraoQXMsduOsJg
- [wiki/Crypto_Wars#Front_doors](https://en.wikipedia.org/wiki/Crypto_Wars#Front_doors), [wiki/Crypto_Wars#2015_UK_call_for_outlawing_non-backdoored_cryptography](https://en.wikipedia.org/wiki/Crypto_Wars#2015_UK_call_for_outlawing_non-backdoored_cryptography)
- https://zh.wikipedia.org/wiki/%E5%AF%86%E7%A0%81%E6%88%98%E4%BA%89

**Reply by emmansun** [reply]
说得好听一点是强监管
